include/shark/Algorithms/Trainers/CSvmTrainer.h

```diff
@@ -266,9 +266,22 @@ class CSvmTrainer : public AbstractSvmTrainer<
 			typedef SvmShrinkingProblem<SVMProblemType> ProblemType;
 			ProblemType problem(svmProblem,base_type::m_shrinking);
 			QpSolver< ProblemType > solver(problem);
-			if (svm.alpha().size1() > 0 && svm.alpha().size2() == 1) {
-				problem.setInitialSolution(blas::column(svm.alpha(), 0));
+			// truncate the existing solution to the bounds
+			RealVector const& reg = this->regularizationParameters();
+			double C_minus = reg(0);
+			double C_plus = (reg.size() == 1) ? reg(0) : reg(1);
+			for (std::size_t i=0, b=0; b<dataset.labels().numberOfBatches(); b++)
```

Comment:
you can make your life easier by iterating over dataset.labels().elements() and thus can skip the for loop over the batches. use autop for the iterator type then, or for(auto label: dataset.labels().elements()) and increment an additional counter.

also: i do not see where you change i in this loop!